### PR TITLE
Extended syntax: Named alternatives, tests

### PR DESCRIPTION
**Pull request by Anabra**
See #32

### CODE REVIEW

**Comment by Anabra** [comment]
This should fix the failing test cases.

File: grin/test/AbstractInterpretation/ExtendedSyntax/IRSpec.hs [hunk 1]
```diff
@@ -42,42 +41,102 @@ randomizeInstructions is0 = do
...
-               t2 <- store (CInt 10000)
-               t3 <- store (Fupto t1 t2)
-               t4 <- store (Fsum t3)
-               (CInt r') <- eval t4
-               _prim_int_print r'
+testProgram = [prog|
+  grinMain =
+    v.0 <- do
+      y.0 <- pure 1
+      pure (CInt y.0)
```
Comment (csabahruska) [inline grin/test/AbstractInterpretation/ExtendedSyntax/IRSpec.hs]: why `do`? why isn't it bind normalised?

File: grin/test/AbstractInterpretation/ExtendedSyntax/OptimiseAbstractProgramSpec.hs [hunk 2]
```diff
@@ -41,42 +40,102 @@ spec = do
...
-               t3 <- store (Fupto t1 t2)
-               t4 <- store (Fsum t3)
-               (CInt r') <- eval t4
-               _prim_int_print r'
+testProgram = withPrimPrelude [prog|
+  grinMain =
+    v.0 <- do
+      y.0 <- pure 1
+      pure (CInt y.0)
+    t1 <- store v.0
```
Comment (csabahruska) [inline grin/test/AbstractInterpretation/ExtendedSyntax/OptimiseAbstractProgramSpec.hs]: bind normalisation

File: grin/test/AbstractInterpretation/ExtendedSyntax/SharingSpec.hs [hunk 3]
```diff
@@ -107,43 +102,103 @@ calcSharingResult prog
...
-                               pure w
-              (Fsum c)      -> z <- sum c
-                               update q z
-                               pure z
+-- The syntax conversion preserves the abstarct heap layout
+testProgram = withPrimPrelude [prog|
+  grinMain =
+    v.0 <- do
+      y.0 <- pure 1
+      pure (CInt y.0)
```
Comment (csabahruska) [inline grin/test/AbstractInterpretation/ExtendedSyntax/SharingSpec.hs]: bind normalisation

File: grin/src/Grin/ExtendedSyntax/Parse/AST.hs [hunk 4]
```diff
@@ -49,7 +49,7 @@ ifThenElse i = do
 
 simpleExp :: Pos -> Parser SimpleExp
 simpleExp i = SReturn <$ kw "pure" <*> value <|>
-              ECase <$ kw "case" <*> var <* kw "of" <*> (L.indentGuard sc GT i >>= some . (\pos -> try (alternative pos) <|> nAlternative pos)) <|>
+              ECase <$ kw "case" <*> var <* kw "of" <*> (L.indentGuard sc GT i >>= some . nAlternative) <|>
```
Comment (andorp) [inline grin/src/Grin/ExtendedSyntax/Parse/AST.hs]: From now on, only named alternatives are possible? Am I right?

File: grin/src/Grin/ExtendedSyntax/Pretty.hs [hunk 5]
```diff
@@ -146,7 +146,7 @@ instance Pretty Lit where
 instance Pretty BPat where
   pretty = \case
     VarPat v    -> pretty v
-    AsPat v val -> pretty v <> pretty '@' <> pretty '(' <> pretty val <> pretty ')'
+    AsPat v val -> pretty v <> pretty '@' <> pretty val
```
Comment (andorp) [inline grin/src/Grin/ExtendedSyntax/Pretty.hs]: Minor: Calling it 'val' is a bit misleading here.

File: grin/src/Transformations/ExtendedSyntax/Names.hs [hunk 6]
```diff
@@ -128,6 +128,7 @@ mapNameDefExpM f = \case
   EBind leftExp (VarPat var) rightExp -> do EBind leftExp <$> (VarPat <$> f var) <*> pure rightExp
   EBind leftExp (AsPat var val) rightExp -> EBind leftExp <$> (AsPat <$> f var <*> mapNamesValM f val) <*> pure rightExp
   Alt cpat body               -> Alt <$> mapNamesCPatM f cpat <*> pure body
```
Comment (andorp) [inline grin/src/Transformations/ExtendedSyntax/Names.hs]: Is there any use-case left for the Alt? Should we remove it soon?

File: grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs [hunk 7]
```diff
@@ -112,10 +112,12 @@ spec = do
             f x =
               z1 <- pure 0
               case x of
-                0 -> x0 <- pure (CInt z1)
-                     pure x0
-                1 -> x1 <- pure (CBool z1)
-                     pure x1
+                0@_1 ->
```
Comment (andorp) [inline grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs]: OMG :) This is soo confusing, first I parsed it as a numeral literal... :D

Comment (andorp) [reply grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs]: You should have called these as "a1" and so on...

File: grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs [hunk 8]
```diff
@@ -526,11 +570,11 @@ spec = do
...
 
                 p0 <- case z0 of
-                  0 -> store z1
-                  1 -> pure (#undefined :: #ptr)
+                  0@_1 -> store z1
+                  1@_2 -> pure (#undefined :: #ptr)
                 n0 <- fetch p0
                 n1 <- pure (#undefined :: {CNode[#ptr]})
-                _1@(CNode p1) <- pure n1
+                _3@(CNode p1) <- pure n1
```
Comment (andorp) [inline grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs]: This is clearly confusing. We should stick with one @ notation. Value/Pattern @ Var or Var @ Value/Pattern

File: grin/test/Transformations/ExtendedSyntax/MangleNamesSpec.hs [hunk 9]
```diff
@@ -32,22 +32,39 @@ spec = do
...
+      grinMain =
+        n0 <- pure 5
+        n1@n2 <- pure n0
+        v@(CInt k) <- pure (CInt n0)
+        pure k
+      |]
+    let after = [prog|
+      name.0 =
+        name.1 <- pure 5
+        name.2@name.3 <- pure name.1
```
Comment (andorp) [inline grin/test/Transformations/ExtendedSyntax/MangleNamesSpec.hs]: I think this is very confusing. I agree that the Node Binding should have a unique name via the as pattern. But I see only complication having an as pattern for a variable.

File: grin/test/AbstractInterpretation/ExtendedSyntax/IRSpec.hs (hunk 1 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/IRSpec.hs]: I forgot a bind normalisation in the conversion function (old AST -> new AST).

Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/IRSpec.hs]: Done.

File: grin/test/AbstractInterpretation/ExtendedSyntax/OptimiseAbstractProgramSpec.hs (hunk 2 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/OptimiseAbstractProgramSpec.hs]: Done.

File: grin/test/AbstractInterpretation/ExtendedSyntax/SharingSpec.hs (hunk 3 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/SharingSpec.hs]: Done.

File: grin/src/Grin/ExtendedSyntax/Parse/AST.hs (hunk 4 above)
Comment (Anabra) [reply grin/src/Grin/ExtendedSyntax/Parse/AST.hs]: Yes. From now on, the Parser only recognizes named alternatives.

File: grin/src/Transformations/ExtendedSyntax/Names.hs (hunk 6 above)
Comment (Anabra) [reply grin/src/Transformations/ExtendedSyntax/Names.hs]: Yes, we should remove it in the next PR.

Comment (Anabra) [reply grin/src/Transformations/ExtendedSyntax/Names.hs]: I already removed it on another branch: [32-alt-remove-unnamed](https://github.com/grin-compiler/grin/tree/32-alt-remove-unnamed)

File: grin/test/Transformations/ExtendedSyntax/MangleNamesSpec.hs (hunk 9 above)
Comment (Anabra) [reply grin/test/Transformations/ExtendedSyntax/MangleNamesSpec.hs]: We will restrict the available patterns only to nodes in the future. I added a few lines about this to #32.

**Comment by Anabra** [comment]
We will restrict the available patterns only to nodes in the future. I added a few lines about this to #32.

File: grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs (hunk 8 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs]: Alright, then we shall swap the variable name and the pattern for as-patterns. I added a few lines about this to #32.

File: grin/src/Transformations/ExtendedSyntax/Names.hs (hunk 6 above)
Comment (Anabra) [reply grin/src/Transformations/ExtendedSyntax/Names.hs]: I already removed it on another branch: [32-alt-remove-unnamed](https://github.com/grin-compiler/grin/tree/32-alt-remove-unnamed)

File: grin/src/Grin/ExtendedSyntax/Pretty.hs (hunk 5 above)
Comment (Anabra) [reply grin/src/Grin/ExtendedSyntax/Pretty.hs]: Renamed it to `pat`.

File: grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs (hunk 7 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs]: I added spaces around the `@`.

Comment (csabahruska) [inline grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs]: Put spaces around `@`.

File: grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs [hunk 10]
```diff
@@ -176,8 +180,9 @@ spec = do
               p0 <- store n0
               n1 <- pure (CCons z0 p0)
               case n1 of
-                (CCons x pxs) -> xs <- fetch pxs
-                                 pure 5
+                (CCons x pxs)@_1 ->
```
Comment (csabahruska) [inline grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs]: put spaces around `@`

File: grin/test/AbstractInterpretation/ExtendedSyntax/HptSpec.hs [hunk 11]
```diff
@@ -49,7 +49,7 @@ spec = do
             grinMain =
               k0 <- pure 0
               x0 <- pure (CInt k0)
-              x1@x2 <- pure x0
+              x2@x1 <- pure x0
```
Comment (csabahruska) [inline grin/test/AbstractInterpretation/ExtendedSyntax/HptSpec.hs]: add spaces around `@`

File: grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs [hunk 12]
```diff
@@ -31,6 +31,9 @@ import AbstractInterpretation.ExtendedSyntax.EffectTrackingSpec (calcEffects)
 
    Variables with names like "_<i>" are introduced just for named bindings.
    They will never be used.
+
+   Variables with names like "alt<i>" are introduced just for named case alternatives.
+   They will never be used.
```
Comment (csabahruska) [inline grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs]: They will never be used (yet)

File: grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs [hunk 13]
```diff
@@ -638,9 +699,9 @@ spec = describe "Live Variable Analysis" $ do
             z3 <- pure 0
             n <- fetch p0
             x <- case n of
-              (CNil c0) -> pure z3
-              (COne c1) -> pure z3
-              #default  -> pure z3
+              (CNil c0)@alt0 -> pure z3
+              (COne c1)@alt1 -> pure z3
+              #default@alt2  -> pure z3
```
Comment (csabahruska) [inline grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs]: add spaces around `@`

File: grin/test/AbstractInterpretation/ExtendedSyntax/OptimiseAbstractProgramSpec.hs [hunk 14]
```diff
@@ -41,42 +40,92 @@ spec = do
...
-               _prim_int_print r'
+testProgram = withPrimPrelude [prog|
+  grinMain =
+    y.0 <- pure 1
+    v.0 <- pure (CInt y.0)
+    t1 <- store v.0
+    y.1 <- pure 10000
+    v.1 <- pure (CInt y.1)
+    t2 <- store v.1
+    y.3 <- pure t2
```
Comment (csabahruska) [inline grin/test/AbstractInterpretation/ExtendedSyntax/OptimiseAbstractProgramSpec.hs]: Do not rename variables. Remove `var2 <- pure var1` lines. Apply copy propagation pass.

File: grin/test/AbstractInterpretation/ExtendedSyntax/SharingSpec.hs [hunk 15]
```diff
@@ -107,43 +102,93 @@ calcSharingResult prog
...
+-- The syntax conversion preserves the abstarct heap layout
+testProgram = withPrimPrelude [prog|
+  grinMain =
+    y.0 <- pure 1
+    v.0 <- pure (CInt y.0)
+    t1 <- store v.0
+    y.1 <- pure 10000
+    v.1 <- pure (CInt y.1)
+    t2 <- store v.1
+    y.3 <- pure t2
```
Comment (csabahruska) [inline grin/test/AbstractInterpretation/ExtendedSyntax/SharingSpec.hs]: apply copy propagation.

File: grin/test/ExtendedSyntax/LintSpec.hs [hunk 16]
```diff
@@ -76,7 +76,7 @@ spec = do
           main =
             s <- pure #"string"
             case s of
-              #"string" -> pure ()
+              #"string"@_1 -> pure ()
```
Comment (csabahruska) [inline grin/test/ExtendedSyntax/LintSpec.hs]: add spaces around `@`.

File: grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs (hunk 7 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs]: Done

File: grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs (hunk 10 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs]: Done

File: grin/test/AbstractInterpretation/ExtendedSyntax/HptSpec.hs (hunk 11 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/HptSpec.hs]: Done

File: grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs (hunk 12 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs]: This is just a naming convention in the tests.

File: grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs (hunk 13 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs]: Done

File: grin/test/AbstractInterpretation/ExtendedSyntax/OptimiseAbstractProgramSpec.hs (hunk 14 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/OptimiseAbstractProgramSpec.hs]: Done

File: grin/test/AbstractInterpretation/ExtendedSyntax/SharingSpec.hs (hunk 15 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/SharingSpec.hs]: Done

File: grin/test/ExtendedSyntax/LintSpec.hs (hunk 16 above)
Comment (Anabra) [reply grin/test/ExtendedSyntax/LintSpec.hs]: Done

File: grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs [hunk 17]
```diff
@@ -424,7 +468,7 @@ spec = do
       let exp = [prog|
             grinMain =
                 z0 <- pure 0
-                a1@a2 <- pure (CInt z0)
+                a2@a1 <- pure (CInt z0)
```
Comment (csabahruska) [inline grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs]: space around `@`? :D

File: grin/src/Transformations/ExtendedSyntax/Conversion.hs [hunk 18]
```diff
@@ -257,14 +262,27 @@ convertToNew :: Exp -> New.Exp
...
+  . bindNormalisation
+  . appArgumentNaming
+  . bindNormalisation
+  . fst . bindingPatternSimplification
+  . bindNormalisation
+  . fst . producerNameIntroduction
+  . bindNormalisation
+
+  where
+    -- SDVE that infers the type env and the effect map
```
Comment (andorp) [inline grin/src/Transformations/ExtendedSyntax/Conversion.hs]: Why do we need a Simple Dead Variable Elimination for the renaming? This looks suspicious :)

File: grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs [hunk 19]
```diff
@@ -747,16 +814,16 @@ spec = describe "Live Variable Analysis" $ do
             n0 <- pure (CBool z0)
             n1 <- pure (CWord z0)
             p0 <- case z0 of
-                    0 -> store n0
-                    1 -> store n1
+              0@alt0 -> store n0
```
Comment (andorp) [inline grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs]: Do we want to have extra spaces here too?

Comment (andorp) [reply grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs]: And at all the subsequent instances.

File: grin/src/Transformations/ExtendedSyntax/Conversion.hs (hunk 18 above)
Comment (Anabra) [reply grin/src/Transformations/ExtendedSyntax/Conversion.hs]: See the comment in copy propagation. CP does not remove the resulting dead bindings. IMO, this should be fixed in the future.

File: grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs (hunk 19 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/LiveVariableSpec.hs]: Yes, I will put spaces here too.

File: grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs (hunk 17 above)
Comment (Anabra) [reply grin/test/AbstractInterpretation/ExtendedSyntax/CreatedBySpec.hs]: Done

**Pull request merged**